components/supervisor/pkg/ports/ports-config.go

```diff
@@ -0,0 +1,226 @@
+		if rangeConfig.Start <= port && port <= rangeConfig.End {
+			return &gitpod.PortConfig{
+				Port:       float64(port),
+				OnOpen:     rangeConfig.OnOpen,
+				Visibility: rangeConfig.Visibility,
+			}, true
+		}
+	}
+	return nil, false
+}
+
+// ConfigInterace allows to watch port configurations
+type ConfigInterace interface {
+	// Observe provides channels triggered whenever the port configurations are changed.
+	Observe(ctx context.Context) (<-chan *Configs, <-chan error)
+}
+
+// ConfigService allows to watch port configurations
+type ConfigService struct {
+	workspaceID   string
+	configService gitpod.ConfigInterface
+	gitpodAPI     gitpod.APIInterface
+	parser        *configParser
+}
+
+// NewConfigService creates a new instance of ConfigService
+func NewConfigService(workspaceID string, configService gitpod.ConfigInterface, gitpodAPI gitpod.APIInterface) *ConfigService {
+	return &ConfigService{
+		workspaceID:   workspaceID,
+		configService: configService,
+		gitpodAPI:     gitpodAPI,
+		parser:        newConfigParser(),
+	}
+}
+
+// Observe provides channels triggered whenever the port configurations are changed.
+func (service *ConfigService) Observe(ctx context.Context) (<-chan *Configs, <-chan error) {
+	updatesChan := make(chan *Configs)
+	errorsChan := make(chan error, 1)
+
+	go func() {
+		defer close(updatesChan)
+		defer close(errorsChan)
+
+		configs, errs := service.configService.Observe(ctx)
+
+		current := &Configs{}
+		if service.gitpodAPI != nil {
+			info, err := service.gitpodAPI.GetWorkspace(ctx, service.workspaceID)
+			if err != nil {
+				errorsChan <- err
+			} else {
+				current.workspaceConfigs = service.parser.parseWorkspaceConfigs(info.Workspace.Config.Ports)
+				updatesChan <- &Configs{workspaceConfigs: current.workspaceConfigs}
+			}
+		} else {
+			errorsChan <- errors.New("could not connect to Gitpod API to fetch workspace port configs")
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-errs:
+				errorsChan <- err
+			case config := <-configs:
+				if service.update(config, current) {
+					updatesChan <- &Configs{
+						workspaceConfigs:     current.workspaceConfigs,
+						instancePortConfigs:  current.instancePortConfigs,
+						instanceRangeConfigs: current.instanceRangeConfigs,
+					}
+				}
+			}
+		}
+	}()
+	return updatesChan, errorsChan
+}
+
+func (service *ConfigService) update(config *gitpod.GitpodConfig, current *Configs) bool {
+	currentPortConfigs, currentRangeConfigs := current.instancePortConfigs, current.instanceRangeConfigs
+	var ports []*gitpod.PortsItems
+	if config != nil {
+		ports = config.Ports
+	}
+	portConfigs, rangeConfigs := service.parser.parseInstanceConfigs(ports)
+	current.instancePortConfigs = portConfigs
+	current.instanceRangeConfigs = rangeConfigs
+	return !(cmp.Equal(currentPortConfigs, portConfigs, cmpopts.SortMaps(func(x, y uint32) bool { return x < y })) && cmp.Equal(currentRangeConfigs, rangeConfigs))
```

Comment:
go-cmp is primarily meant for testing. The `reflect` package has a `DeepEqual` function that does what you need it to do.
Note: maps are inherently unordered, and `DeepEqual` does not care about the order. See
  - https://golang.org/pkg/reflect/#DeepEqual
  - https://play.golang.org/p/7UoLMKkt9QX


```suggestion
return !reflect.DeepEqual(currentPortConfigs, portConfigs) || !reflect.DeepEqual(currentRangeConfigs, rangeConfigs)
```